prov/usnic/src/usdf_cm.c

```diff
@@ -209,25 +209,82 @@ usdf_cm_msg_accept(struct fid_ep *fep, const void *param, size_t paramlen)
 	return ret;
 }
 
-/*
- * Connection request attempt failed
+/* Given a connection request structure containing data, make a copy of the data
+ * that can be accessed in error entries on the EQ. The return value is the size
+ * of the data stored in the error entry. If the return value is a non-negative
+ * value, then the function has suceeded and the size and output data can be
+ * assumed to be valid. If the function fails, then the data will be NULL and
+ * the size will be a negative error value.
  */
-void
-usdf_cm_msg_connreq_failed(struct usdf_connreq *crp, int error)
+static int usdf_cm_generate_err_data(struct usdf_eq *eq,
+		struct usdf_connreq *crp, void **data)
 {
 	struct usdf_err_data_entry *err_data_entry;
 	struct usdf_connreq_msg *reqp;
+	size_t entry_size;
+	size_t data_size;
+
+	if (!eq || !crp || !data) {
+		USDF_DBG_SYS(EP_CTRL,
+				"eq, crp, or data is NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	/* Initialize to NULL so data can't be used in the error case. */
+	*data = NULL;
+
+	reqp = (struct usdf_connreq_msg *) crp->cr_data;
+
+	/* This is a normal case, maybe there was no data. */
+	if (!reqp || !reqp->creq_datalen)
+		return 0;
+
+	data_size = reqp->creq_datalen;
+
+	entry_size = sizeof(*err_data_entry) + data_size;
+
+	err_data_entry = calloc(1, entry_size);
+	if (!err_data_entry) {
+		USDF_WARN_SYS(EP_CTRL,
+				"failed to allocate err data entry\n");
+		return -FI_ENOMEM;
+	}
+
+	/* This data should be copied and owned by the provider. Keep
+	 * track of it in the EQ, this will be freed in the next EQ read
+	 * call after it has been read.
+	 */
+	memcpy(err_data_entry->err_data, reqp->creq_data, data_size);
+	slist_insert_tail(&err_data_entry->entry, &eq->eq_err_data);
+
+	*data = err_data_entry->err_data;
+
+	return data_size;
+}
+
+/* Report a connection management related failure. Sometimes the data should be
+ * skipped, in this case the skip data argument can be set to something with a
+ * truthy value.
+ *
+ * If data is to be generated for the error entry, then the connection request
+ * is assumed to have the data size in host order. If something fails during
+ * processing of the error data, then the EQ entry will still be generated
+ * without the error data.
+ */
+void usdf_cm_report_failure(struct usdf_connreq *crp, int error, bool skip_data)
```

Comment:
Minor nit: I'm not a big fan of parameters that exhibit negative logic (e.g., "if `skip_data` is `true`, then you *don't* do something").  If you care, it would be nice to switch to a parameter that exhibits positive logic.